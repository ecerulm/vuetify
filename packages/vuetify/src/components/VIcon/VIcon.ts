import './VIcon.sass'

// Utilities
import { computed, defineComponent, h } from 'vue'
import { useIcon, makeSizeProps, useSizeClass } from '@/composables'
import makeProps from '@/util/makeProps'

// Types
import type { PropType } from 'vue'
import type { VuetifyIcon } from '@/composables'

export const VSvgIcon = defineComponent({
  name: 'VSvgIcon',
  inheritAttrs: false,
  props: {
    icon: {
      type: String,
      required: true,
    },
    tag: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      required: true,
    },
  },
  setup (props, context) {
    return () => h(props.tag, {
      ...context.attrs,
      style: null,
    }, [
      h('svg', {
        style: context.attrs.style,
        class: 'v-icon__svg',
        xmlns: 'http://www.w3.org/2000/svg',
        viewBox: '0 0 24 24',
        role: 'img',
        'aria-hidden': true,
      }, [h('path', { d: props.icon })]),
    ])
  },
})

export const VLigatureIcon = defineComponent({
  name: 'VLigatureIcon',
  props: {
    icon: {
      type: String,
      required: true,
    },
    tag: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      required: true,
    },
  },
  setup (props) {
    return () => h(props.tag, {
      class: props.type,
    }, [props.icon])
  },
})

export const VClassIcon = defineComponent({
  name: 'VClassIcon',
  props: {
    icon: {
      type: String,
      required: true,
    },
    tag: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      required: true,
    },
  },
  setup (props) {
    return () => h(props.tag, {
      class: [props.type, props.icon],
    })
  },
})

export default defineComponent({
  name: 'VIcon',

  props: makeProps({
    disabled: Boolean,
    left: Boolean,
    right: Boolean,
    tag: {
      type: String,
      required: false,
      default: 'i',
    },
    icon: {
      type: [String, Object] as PropType<VuetifyIcon>,
      required: true,
    },
    type: {
      type: String,
      default: 'mdi',
    },
    ...makeSizeProps(),
  }),

  setup (props, context) {
    const { sizeClass } = useSizeClass(props)
    const { icon } = useIcon(props)

    const styles = computed(() => !sizeClass.value ? ({
      'font-size': props.size,
      width: props.size,
      height: props.size,
    }) : null)

    const hasClickListener = !!context.attrs.onClick
    const tag = hasClickListener ? 'button' : props.tag

    const classes = computed(() => ([
      'v-icon',
      'notranslate',
      {
        'v-icon--disabled': props.disabled,
        'v-icon--left': props.left,
        'v-icon--right': props.right,
        'v-icon--link': hasClickListener,
      },
      sizeClass.value,
    ]))

    return () => {
      return icon.value.component({
        ...context.attrs,
        tag,
        type: props.type,
        icon: icon.value.icon,
        class: classes.value,
        style: styles.value,
      })
    }
  },
})
