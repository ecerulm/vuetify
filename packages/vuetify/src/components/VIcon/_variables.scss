@import '../../styles/styles.sass';

$icon-size: 24px !default;
$icon-size-dense: 20px !default;

$icon-sizes: () !default;
$icon-sizes: map-deep-merge(
  (
    'x-small': 16px,
    'small': 20px,
    'default': 24px,
    'large': 28px,
    'x-large': 32px,
  ),
  $icon-sizes
);
