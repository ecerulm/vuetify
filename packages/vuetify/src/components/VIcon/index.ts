import VIcon from './VIcon'

export { VLigatureIcon, VClassIcon, VSvgIcon } from './VIcon'
export { VIcon }
export default VIcon
