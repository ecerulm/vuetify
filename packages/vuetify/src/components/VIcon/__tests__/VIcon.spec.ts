// Components
import VIcon from '../VIcon'

// Utilities
import { defineComponent, h } from 'vue'
import { mount } from '@vue/test-utils'
import * as framework from '@/framework'
import type { VuetifyIcons } from '@/composables'

describe('VIcon', () => {
  let vuetifyMock: jest.SpyInstance

  beforeEach(() => {
    vuetifyMock = jest.spyOn(framework, 'useVuetify').mockImplementation(() => ({
      defaults: {
        global: {},
      },
      icon: {
        icons: {
          checkboxOn: 'check_box',
          checkboxOff: {
            component: defineComponent({
              props: {
                foo: String,
              },
              setup (props) {
                return () => h('div', [props.foo])
              },
            }),
          },
        } as unknown as VuetifyIcons,
      },
    }))
  })

  afterEach(() => {
    vuetifyMock.mockClear()
  })

  it('should render component', () => {
    const wrapper = mount(VIcon, {
      props: {
        icon: 'add',
      },
    })

    expect(wrapper.text()).toBe('add')
    expect(wrapper.element.className).toBe('v-icon notranslate material-icons theme--light')
  })

  // it('should render a colored component', () => {
  //   const wrapper = mount(VIcon, { props: { color: 'green lighten-1' } })

  //   expect(wrapper.element.classList).toContain('green--text')
  //   expect(wrapper.element.classList).toContain('text--lighten-1')
  // })

  it('should render a disabled component', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add', disabled: true } })

    expect(wrapper.element.classList).toContain('v-icon--disabled')
  })

  it('should use default size if none provided', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add' } })

    expect(wrapper.element.classList).toContain('v-size--default')
  })

  it('should use provided size', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add', size: 'x-small' } })

    expect(wrapper.element.classList).toContain('v-size--x-small')
  })

  // it('should render a specific size with String type', () => {
  //   const wrapper = mountFunction({ props: { size: '112px' } })

  //   expect(wrapper.element.style.fontSize).toBe('112px')
  // })

  // it('should render a specific size with Number type', () => {
  //   const wrapper = mountFunction({ props: { size: '112' } })

  //   expect(wrapper.element.style.fontSize).toBe('112px')
  // })

  it('should render a left aligned component', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add', left: true } })

    expect(wrapper.element.classList).toContain('v-icon--left')
  })

  it('should render a right aligned component', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add', right: true } })

    expect(wrapper.element.classList).toContain('v-icon--right')
  })

  it('should render a component with aria-hidden attr', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add' }, attrs: { 'aria-hidden': 'foo' } })

    expect(wrapper.element.getAttribute('aria-hidden')).toBe('foo')
  })

  it('should allow third-party icons when using <icon>- prefix', () => {
    const wrapper = mount(VIcon, { props: { icon: 'fa-add' } })

    expect(wrapper.text()).toBe('')
    expect(wrapper.element.classList).toContain('fa')
    expect(wrapper.element.classList).toContain('fa-add')
  })

  it('should support font awesome 5 icons when using <icon>- prefix', () => {
    const wrapper = mount(VIcon, { props: { icon: 'fab fa-facebook' } })

    expect(wrapper.text()).toBe('')
    expect(wrapper.element.classList).toContain('fab')
    expect(wrapper.element.classList).toContain('fa-facebook')
  })

  // it('should allow the use of v-text', () => {
  //   const wrapper = mountFunction({
  //     domProps: { textContent: 'fa-home' },
  //   })

  //   expect(wrapper.text()).toBe('')
  //   expect(wrapper.element.className).toBe('v-icon notranslate fa fa-home theme--light')
  // })

  // it('should allow the use of v-html', () => {
  //   const wrapper = mountFunction({
  //     domProps: { innerHTML: 'fa-home' },
  //   })

  //   expect(wrapper.text()).toBe('')
  //   expect(wrapper.element.className).toBe('v-icon notranslate fa fa-home theme--light')
  // })

  // it('set font size from helper prop', async () => {
  //   const iconFactory = size => mountFunction({
  //     props: { [size]: true },
  //   })

  //   const small = iconFactory('small')
  //   expect(small.html()).toMatchSnapshot()

  //   const medium = iconFactory('medium')
  //   expect(medium.html()).toMatchSnapshot()

  //   const large = iconFactory('large')
  //   expect(large.html()).toMatchSnapshot()

  //   const xLarge = iconFactory('xLarge')
  //   expect(xLarge.html()).toMatchSnapshot()
  // })

  // it('should have proper classname', () => {
  //   const wrapper = mountFunction({
  //     props: {
  //       color: 'primary',
  //     },
  //     domProps: {
  //       innerHTML: 'fa-lock',
  //     },
  //   })

  //   expect(wrapper.element.className).toBe('v-icon notranslate fa fa-lock theme--light primary--text')
  // })

  it('should render globally defined icon', () => {
    const wrapper = mount(VIcon, { props: { icon: '$checkboxOn' } })

    expect(wrapper.text()).toBe('check_box')
    expect(wrapper.element.className).toBe('v-icon notranslate material-icons theme--light')
  })

  //   it('should render MD left icon from $prev', () => {
  //     const wrapper = mountFunction({}, '$prev')

  //     expect(wrapper.text()).toBe('chevron_left')
  //     expect(wrapper.element.className).toBe('v-icon notranslate material-icons theme--light')
  //   })
  // })

  it('should use an <i> tag if none provided', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add' } })

    expect(wrapper.element.localName).toBe('i')
  })

  it('sets tag from from prop if provided', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add', tag: 'span' } })

    expect(wrapper.element.localName).toBe('span')
  })

  it('should render custom component', () => {
    const wrapper = mount(VIcon, { props: { icon: '$checkboxOff' } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  //   it('should render a colored component', () => {
  //     const wrapper = mountFunction({ props: { color: 'green lighten-1' } }, '$testIcon')

  //     expect(wrapper.element.classList).toContain('green--text')
  //     expect(wrapper.element.classList).toContain('text--lighten-1')
  //   })

  it('should render a disabled icon', () => {
    const wrapper = mount(VIcon, { props: { icon: 'add', disabled: true } })

    expect(wrapper.element.classList).toContain('v-icon--disabled')
  })

  //   it('should set font size from helper prop', async () => {
  //     const iconFactory = size => mountFunction({
  //       props: { [size]: true },
  //     }, '$testIcon')

  //     const small = iconFactory('small')
  //     expect(small.html()).toMatchSnapshot()

  //     const medium = iconFactory('medium')
  //     expect(medium.html()).toMatchSnapshot()

  //     const large = iconFactory('large')
  //     expect(large.html()).toMatchSnapshot()

  //     const xLarge = iconFactory('xLarge')
  //     expect(xLarge.html()).toMatchSnapshot()
  //   })

  //   it('should render a left aligned component', () => {
  //     const wrapper = mountFunction({ props: { left: true } }, '$testIcon')

  //     expect(wrapper.element.classList).toContain('v-icon--left')
  //   })

  //   it('should render a right aligned component', () => {
  //     const wrapper = mountFunction({ props: { right: true } }, '$testIcon')

  //     expect(wrapper.element.classList).toContain('v-icon--right')
  //   })

  it('should render clickable icon if click handler specified', () => {
    const clickHandler = jest.fn()
    const wrapper = mount(VIcon, { props: { icon: 'add', onClick: clickHandler } })
    wrapper.trigger('click')

    expect(wrapper.element.classList).toContain('v-icon--link')
    expect(clickHandler).toHaveBeenCalled()
    expect(wrapper.element.getAttribute('aria-hidden')).toBe('false')
    expect(wrapper.element.getAttribute('type')).toBe('button')
  })

  //   it('should trim name', () => {
  //     const wrapper = mountFunction({}, ' add ')

  //     expect(wrapper.text()).toBe('add')
  //   })

  it('should render an svg icon', async () => {
    const wrapper = mount(VIcon, { props: { icon: 'M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z' } })

    expect(wrapper.html()).toMatchSnapshot()
  })
  // })
})
