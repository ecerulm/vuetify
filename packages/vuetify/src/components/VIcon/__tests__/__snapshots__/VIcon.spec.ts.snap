// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VIcon should render an svg icon 1`] = `
<i class="v-icon notranslate  v-size--default M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
   aria-hidden="true"
   disabled="false"
>
  <svg class="v-icon__svg"
       xmlns="http://www.w3.org/2000/svg"
       viewbox="0 0 24 24"
       role="img"
       aria-hidden="true"
  >
    <path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z">
    </path>
  </svg>
</i>
`;

exports[`VIcon should render custom component 1`] = `
<i class="v-icon notranslate  v-size--default"
   aria-hidden="true"
   disabled="false"
>
  <div class="v-icon__component">
    bar
  </div>
</i>
`;
